src/Hedgehog/Gen.fs

```diff
@@ -515,21 +515,43 @@ module Gen =
         return System.Guid bs
     }
 
-    /// Generates a random instant in time expressed as a date and time of day.
-    [<CompiledName("DateTime")>]
-    let dateTime : Gen<System.DateTime> =
-        let minTicks =
-            System.DateTime.MinValue.Ticks
-        let maxTicks =
-            System.DateTime.MaxValue.Ticks
+    /// Generates a random DateTime using the specified range.
+    [<CompiledName("DateTimeRanged")>]
+    let dateTimeRanged (range : Range<DateTime>) : Gen<System.DateTime> =
         gen {
-            let! ticks =
-                Range.constantFrom
-                    (System.DateTime (2000, 1, 1)).Ticks minTicks maxTicks
-                |> integral
+            let! ticks = range |> Range.map (fun dt -> dt.Ticks) |> integral
             return System.DateTime ticks
         }
 
+    /// Generates a random DateTimeOffset using the specified range.
+    [<CompiledName("DateTimeOffsetRanged")>]
+    let dateTimeOffsetRanged (range : Range<DateTimeOffset>) : Gen<System.DateTimeOffset> =
+        gen {
+            let! ticks = range |> Range.map (fun dt -> dt.Ticks) |> integral
+            // Ensure there is no overflow near the edges when adding the offset
+            let minOffsetMinutes =
+              max 
+                (-14L * 60L)
+                ((DateTimeOffset.MaxValue.Ticks - ticks) / TimeSpan.TicksPerMinute * -1L)
+            let maxOffsetMinutes =
+              min 
+                (14L * 60L)
+                ((ticks - DateTimeOffset.MinValue.Ticks) / TimeSpan.TicksPerMinute)
+            let! offsetMinutes = int (Range.linearFrom 0 (Operators.int minOffsetMinutes) (Operators.int maxOffsetMinutes))
+            return System.DateTimeOffset(ticks, TimeSpan.FromMinutes (Operators.float offsetMinutes))
+        }
+
+    /// Generates a random DateTime.
+    [<CompiledName("DateTime")>]
+    let dateTime : Gen<System.DateTime> =
+        dateTimeRanged (Range.constantFrom (DateTime (2000, 1, 1)) DateTime.MinValue DateTime.MaxValue)
```

Comment:
Perhaps we should just deprecate those in favor or the ranged ones with v0.9.0? (That is, we keep the same name `dateTime`, `dateTimeOffset` but we accept `Range` as parameter.)

After all, it's just a one liner to define the old one (assuming we name the ranged one as `dateTime`):

```f#
dateTime (Range.constantFrom (DateTime (2000, 1, 1)) DateTime.MinValue DateTime.MaxValue)
```

I understand that it's a breaking change, but we're v0.x so we should consider that trade-off if it helps keep the API clean.

Reply:
Sure. Do you want me to remove the non-ranged `datetimeoffset` then, since that's a new part of the API? No sense in adding something that's deprecated from day 1. :)

Reply:
Or, when you say "deprecate", do you mean that we should just remove the non-ranged ones altogether?

Reply:
Yes, I think we should take them out. Are you OK with this @TysonMN?

Reply:
If I correctly understand, then yes.  Here are my thoughts based on that linked code.  I think they are consistent with the conclusion being reached here.

There are two functions there: `dateTimeOffsetRanged` and `dateTimeOffset`.  Of course the preferred name is the shorter `dateTimeOffset`.  I think the function accepting `Range<DateTimeOffset>` should have the preferred name (i.e. `dateTimeOffset`).  Then we can either remove the other function or rename it to something like `dateTimeOffsetWithMaxRange`.

Reply:
Yep. I think that's one of the very few (if not the first) case in the last 10+ years where I think introducing a lil breaking change is actually fine.